### ISSUE TITLE
Regroup tests

### ISSUE THREAD

**Issue by larkost** [rethinkdb/rethinkdb#4896]
As part of #3442 I will be re-grouping the tests, as they have organically grown out of the original design. After a brief talk with @danielmewes here is the proposed layout (incomplete, but a working basis):
- server
  - admin
  - changefeeds
  - clustering (including the scenario tests)
  - regression
- client
  - changefeeds
  - connection
  - datum
  - (most/all of `./test/rql_test/src/)
  - regression
- performance
- fuzzing

Any sort of test would be able to go at any point on this hierarchy, so for example YAML-format tests would be perfectly legal in `./test/server` (`test-runner` already allows for this). At the end of #3442 all of the current tests would reside in this same hierarchy.

I will start working on this in the near future, and so would appreciate any improvements or kibitzing on this to come back soon.

**Comment by mlucy** [comment]
For a lot of our tests the line between client and server test is pretty fuzzy.  (Most of the regression tests in `rql_test/src` are really regression tests for server behavior, for example.)  A more natural grouping than "client" might be "query language".

**Comment by larkost** [comment]
We talked over the "client"/"reql" issue, and came down on the "client" side since things like the "connection" tests fall more naturally there. I think most of the regression tests will fall into the "server" side.

**Comment by mlucy** [comment]
Will the polyglot tests be split between client and server tests?

I'm honestly not too sure the distinction works well.  Most of the tests I write test both the client and the server at the same time.  (For instance, when you add a polyglot test for a term, you're adding it both to check that it's implemented in all the clients and to check that the server passes some basic sanity checks on its behavior.)

**Comment by larkost** [comment]
The thinking in that case was that those should probably go in the `client` folder, since you defiantly want to make sure that all of the clients handle it. The `server` folder would be for things where it does not really make sense to test it in all driver languages. I am not really concerned that we wind up getting at least some server testing in on the `client` tests.

**Comment by deontologician** [comment]
Maybe a name like "server_only" and "server_client_interaction" to indicate one set only cares about the server's behavior (and purely as an expedient uses a client to test that behavior), and one set is about testing that the client and server do sane things when interacting

**Comment by mlucy** [comment]
It sort of sounds like the difference is that the `client` section is for tests that exist in all 3 languages and the `server` section is for everything else, is that about right?  Maybe we should just call the client section `polyglot`?

**Comment by larkost** [comment]
@deontologician: I don't think that trying to be that pedantic/literal is really worth it. There is going to be some overlap, and that is ok. These names are only to reflect the over-arching goals of the sections. Additionally, since these are going to be a part of the name you have to type to get to a specific test, making them that long is a bad idea.

@mlucy: I really dislike the name `polyglot`, since that is talking about an implementation detail, not a testing goal. The testing goals really are simple: to test the server and to test the clients. That we are going to have some "chocolate in our peanut butter" is ok. Over time I think people will naturally write tests more aligned with the name, but I anticipate that there will always be some mixed. I don't see that imperfection as being worth obscuring the over-arching goals for.

**Comment by deontologician** [comment]
Ah, I hadn't thought about it being part of the command. In that case
client/server seems fine. Though I think it's not quite accidental
chocolate in the peanut butter since testing the server's interaction is a
part of the goal. but convenience is worth more than that distinction

**Comment by mlucy** [comment]
> The testing goals really are simple: to test the server and to test the clients.

It's true that those are goals, but it's not true that our tests fall cleanly along that line.  It's not like we have mostly chocolate and peanut butter with a little mixing, probably more than half the tests we write now test both the clients and the server.  People will end up having rules in their head like "If I write a test that works in all 3 languages put it in `client`, if I write the same test but only for one language put it in `server`...".  And then when they run up against a novel case, like a test that uses the EventMachine integration to read from lots of feeds, they won't know where to put it.  Is it testing that the EventMachine integration is correct?  Is it testing that having lots of concurrent changefeeds on the server works?  Is it testing that having lots of concurrent changefeeds with the EventMachine integration works?  It's testing all of those!

The choice of which section to put it in is esentially random, and every time you want to find a test you'll have to go look in both.  It doesn't even help reduce the number of tests you have to run, because whenever I make a change to the query language I run all the ReQL tests, so after this divison I'll have to go run both the server tests and the client tests since the current ReQL tests will be split across both.

What I really want from the testing system as a user of it is a way to say something like "I just made a change to the query language, run all the ReQL tests." or "I just made a change to how changefeeds work that shouldn't affect anything else, run just the changefeed section of the ReQL tests" or "I just made a change to how clustering works that doesn't effect query evaluation at all, run all our clustering tests".

Basically, it would be much more useful for me to have tests grouped by what part of the server code they hit.  As a writer of tests, I'll basically be flipping a coin to decide whether a new changefeed test goes in `server/changefeeds` or `client/changefeeds`, and as a user of tests I'm infinitely more likely to want to run `server/changefeeds` and `client/changefeeds` at the same time than to run say `server/changefeeds` and `server/admin` at the same time.

**Comment by mlucy** [comment]
Apologies for the long comment, but it really does seem to me like the primary impact of creating `server/changefeeds` and `client/changefeeds` will just be that I have to look in two places and run two sets of tests.

**Comment by mlucy** [comment]
We talked about this a little more at lunch.  Here's an alternate hierarchy:

* Unit Tests
* Cluster Tests
* ReQL Tests (this has some toplevel tests like `get_all.rb` in addition to the subdirectories below)
  - Changefeed Tests
  - Polyglot Tests (maybe with a different name, but these tests should be in all 3 languages)
* Performance Tests
* Fuzzers

@danielmewes, @deontologician, @VeXocide, @encryptio, @larkost -- do you prefer this hierarchy or the `server/client` one?  (Or one of those but with some minor changes?)

**Comment by deontologician** [comment]
Will the tool transparently handle finding the tests so I don't have to care which directory the tests are in if I don't want to?

**Comment by deontologician** [comment]
To clarify: I think the groups feature is what I will end up using. It'd be nice if we could just specify a group and not have to specify a folder and group. If that's the case, then the hierarchy itself doesn't matter as much to me

**Comment by larkost** [comment]
`group`s as they are implemented in `./test/run` (my starting point) are nearly orthogonal to folder structure. They can be made up of manually created lists, wildcard selections of tests, or even script-driven. I intend to keep that flexibility, while slightly simplifying the implementation/creation.

**Comment by danielmewes** [comment]
I think I prefer @mlucy's hierarchy.

We also need a place for the connection tests. I can see arguments both for putting them in their directory below the top level, or as a subdirectory under "reql". I personally think having them as a top-level folder makes a bit more sense, but either location works for me.
